Implement SpiceCell (#89)
SETI/pds-cspyce: README.md

they can call

```python
cspyce.bodn2c.error.vector(args, ...)
```

> ## RECORD ENHANCEMENTS

Reviewer: ```suggestion
---
## RECORD ENHANCEMENTS
```

Author: Done

Reviewer: I don't understand this fix. You didn't make the change I asked for but seem to have put in a different line in the wrong place.

### SpiceCell
`cspyce` provides an enhanced version of `SpiceCell` to make it simpler to use for
both CSPICE and in Python. 

In the documentation that follows, _active elements_ refer to those elements in the 
`SpiceCell` whose `index` is such that `0 ≤ index < spice_cell.card`. 

#### Constructors
To create a `SpiceCell`:

```python
from cspyce import SpiceCell, SPICE_CELL_INT, SPICE_CELL_DOUBLE

> spice_cell_int = SpiceCell(typeno=SPICE_CELL_INT, size=20)
> spice_cell_double = SpiceCell(typeno=SPICE_CELL_DOUBLE, size=10)

Reviewer: ```suggestion
spice_cell = SpiceCell(typeno=SPICE_CELL_INT, size=20)
spice_cell = SpiceCell(typeno=SPICE_CELL_DOUBLE, size=10)
```

```

With these constructors, `spice_cell.card` will be set to 0 and there will initially be
no active elements.

If you already have data that you want to convert to a spice cell, you can use a simpler
interface:
```python
spice_cell = SpiceCell([1, 2, 3, 4, 5])
spice_cell = SpiceCell(np.arange(1.0, 10.0))
spice_cell = SpiceCell(np.arange(1.0, 10.0), size=40)  # set size explicitly
```
If you use these constructions, Python will create a `SpiceCell` whose size is a little 
bigger than the length of the given data. The passed first argument will be the active
elements.

Any CSPICE function that expects a `SpiceCell` as an input can also be passed an array,
a tuple, or anything that can reasonably be converted into an appropriately typed
`SpiceCell`:

```shell
>>> import cspyce
>>> cspyce.wninsd(2.0, 4.9, (1.0, 3.0, 5.0, 7.0, 9.0, 11.0))
<SpiceCell float 6/12 [ 1.   4.9  5.   7.   9.  11. ]>
```

#### Sequence-like operations

`SpiceCell` offers the following methods so that they behave like Python sequences.

`spice_cell[index]`, `spice_cell[index] = value`

> :  Get or sets the `index`-th element of the `SpiceCell`. 

Reviewer: ```suggestion
:  Gets or sets the `index`-th element of the `SpiceCell`. 
```

   Must have `-len(spice_cell) ≤ index < spice_cell.size`.  As in Python, we allow
   negative indices, but -1, -2, etc count backwards from the last active elements.

`len(spice_cell)`
: Returns the number of active elements of the `SpiceCell`.

`spice_cell.append(value)`
: Appends a new active cell to the end of the `SpiceCell`'s active cells.
The cell's maximum size is grown if necessary. 

`spice_cell.extend(values)`
: Appends all the values of the sequence to the end of the `SpiceCell`'s active cells.
The `SpiceCell`'s maximum size is grown if necessary.

`spice_cell += values`
: Synonym for `spice_cell.extend(values)`

`spice_cell.clear()`
: Make all elements inactive. Same as `spice_cell.card = 0`. 

`iter(spice_cell)` 

> : Iterates through the active elements of `SpiceCell`.

Reviewer: ```suggestion
: Iterates through the active elements of `spice_cell`.
```

> Because of this iterator, `list(spice_cell)`, `set(spice_cell)`, `tuple(spice_cell)` 

Reviewer: ```suggestion
Because of this iterator, `list(spice_cell)`, `set(spice_cell)`, and `tuple(spice_cell)` 
```

work as expected.
You can also write a for loop `for item in spice_cell: ...` to iterate through
the active elements.

> #### SpiceCell specific operations.

Reviewer: ```suggestion
#### SpiceCell-specific operations.
```

`SpiceCell` also has the following methods and properties specific to a `SpiceCell`.

`spice_cell.size`
: Gets the current maximum size of the `SpiceCell`.

`spice_cell.size = value`
: Grows or shrinks the maximum size of the `SpiceCell`.
If the number of active elements is greater than `value`, it is reduced to `value`.

> This operation is a `cspyce` enhancement does not exist in CSPICE.

Reviewer: ```suggestion
This operation is a `cspyce` enhancement that does not exist in CSPICE.
```

`spice_cell.card`

> : Same as `len(spice_cell)`.  (Also known as the "cardinality". Hence the name).

Reviewer: ```suggestion
: Same as `len(spice_cell)`.  (Also known as the "cardinality", hence the name).
```

`spice_cell.card = value`
: Changes the number of active elements in the `SpiceCell`. 
Must have `0 ≤ value < spice_cell.size`.

`spice_cell.as_array()` 

> : Shows the active contents of the array as a numpy array

Reviewer: ```suggestion
: Returns the active contents of the array as a numpy array
```

> `spice_cell.as_intervals` 
> : Shows the active contents of the array as a 2-dimensional `card/2`x 2 array.  

Reviewer: ```suggestion
`spice_cell.as_intervals()` 
: Returns the active contents of the array as a 2-dimensional `card/2`x 2 array.  
```

Author: Mark wanted the word "view" in there, so I changed it to "returns a view of...."

interface:
```python
spice_cell = SpiceCell([1, 2, 3, 4, 5])
spice_cell = SpiceCell(np.arange(1.0, 10.0))
spice_cell = SpiceCell(np.arange(1.0, 10.0), size=40)  # set size explicitly
```
If you use these constructions, Python will create a `SpiceCell` whose size is a little 
bigger than the length of the given data. The passed first argument will be the active
elements.

Any CSPICE function that expects a `SpiceCell` as an input can also be passed an array,
a tuple, or anything that can reasonably be converted into an appropriately typed
`SpiceCell`:

```shell
>>> import cspyce
>>> cspyce.wninsd(2.0, 4.9, (1.0, 3.0, 5.0, 7.0, 9.0, 11.0))
<SpiceCell double 6/12 [ 1.   4.9  5.   7.   9.  11. ]>
```

#### Sequence-like operations

`SpiceCell` offers the following methods so that they behave like Python sequences.

`spice_cell[index]`, `spice_cell[index] = value`

:  Gets or sets the `index`-th element of the `SpiceCell`. 
   Must have `-len(spice_cell) ≤ index < spice_cell.size`.  As in Python, we allow
   negative indices, but -1, -2, etc count backwards from the last active elements.

`len(spice_cell)`

> : Returns the number of active elements of the `SpiceCell`.

Reviewer: I missed this one first time around

> : Returns the number of active elements of the `SpiceCell`.

Reviewer: ```suggestion
: Returns the number of active elements of `spice_cell`.
```

`spice_cell.append(value)`
: If `value` is a float or integer, it is added to the `SpiceCell`'s active cells.
If `value` is an array, all of its elements are added to the `SpiceCell`'s active cells.
The cell's maximum size is grown if necessary. 

`spice_cell.extend(values)`
: Appends each value in `values` to the end of the `SpiceCell`'s active cells.
Each element of `values` can be a float, an integer, or an array.
The `SpiceCell`'s maximum size is grown if necessary.

`spice_cell += values`
: Synonym for `spice_cell.extend(values)`

`spice_cell.clear()`
: Make all elements inactive. Same as `spice_cell.card = 0`. 

`iter(spice_cell)` 
: Iterates through the active elements of `spice_cell`.
Because of this iterator, `list(spice_cell)`, `set(spice_cell)`, and `tuple(spice_cell)` 
work as expected.
You can also write a for loop `for item in spice_cell: ...` to iterate through
the active elements.

> #### SpiceCell-specific operations.

Reviewer: ```suggestion
#### SpiceCell-specific operations
```

`SpiceCell` also has the following methods and properties specific to a `SpiceCell`.

`spice_cell.size`
: Gets the current maximum size of the `SpiceCell`.

`spice_cell.size = value`
: Grows or shrinks the maximum size of the `SpiceCell`.
If the number of active elements is greater than `value`, it is reduced to `value`.
This operation is a `cspyce` enhancement that does not exist in CSPICE.

`spice_cell.card`

> : Same as `len(spice_cell)`.  (Also known as the "cardinality"; hence the name).

Reviewer: ```suggestion
: Same as `len(spice_cell)`.  (Also known as the "cardinality", hence the name).
```

Reviewer: Semicolon is not correct grammar here. A semicolon should start a complete new sentence. The original comma I suggested is correct.

Author: I'm sorry, I'm not going to let Mrs. Harper, my 11th grade English teacher,  fail me even though she's been dead for twenty years.

I typed  - What punctuation to use before "hence" - to Google. 

1. Use a semicolon, not a comma, to join two independent clauses separated by a conjunctive adverb (such as however). The conjunctive adverbs include accordingly, consequently, hence, however, moreover, otherwise, therefore, and thus.

Author: Hmm.  But it's not really an independent clause, so you're probably right.
